### ISSUE TITLE
Intended way to customize CatalogController SearchBuilder behavior?

### ISSUE THREAD

**Issue by jrochkind** [samvera/hyrax#707]
Stock Blacklight generates a `SearchBuilder` class into your app, which by default is used by CatalogController, and you customize the search builder in that generated local class. 

Reasons you might want to customize that includes blacklight_advanced_search or blacklight_range_limit, both of which generate lines into your local `./app/models/search_builder.rb`. 

The intention of the Blacklight architecture is that this locally generated class is your hook for adding custom behavior, whether from a plugin or purely local. 

However, a Sufia/Hyrax app doesn't actually use this for the CatalogController -- the main controller for public searches. Instead, it uses a `[Sufia|Hyrax]::CatalogSearchBuilder`, from gem source I think rather than generated into the local app. 

Sufia hard-codes this for anything using `Sufia::Catalog` module here https://github.com/projecthydra/sufia/blob/818cb4ca53d81fdb37835a0cd7c753d66eacd679/app/controllers/concerns/sufia/catalog.rb

Hyrax is different, and at first blush better, by generating (I think?) a local CatalogController  that sets  the search builder in config: https://github.com/projecthydra-labs/hyrax/blob/544c3d47e2403fffc45f7d986212e053d3913fc1/lib/generators/hyrax/templates/catalog_controller.rb#L17

However, in both cases, still in Hyrax too, it's unclear how I am meant to supply local customization to the SearchBuilder.  The CatalogController SearchBuilder is definitely one that is expected to need local/plugin customization (eg blacklight_range_limit). How does Hyrax intend to do this?  If I manually set my `config.search_builder_class` to something else, so I can customize it, then I don't have the `Hyrax::SearchBuilder` anymore. Unless I inherit from it in my local one, I guess?  That might work, but I'm reluctant to do it without some feedback that it is what Sufia/Hyrax intends. 

[The hyrax docs on SearchBuilders](https://github.com/projecthydra-labs/hyrax/blob/master/app/search_builders/hyrax/README.md)  don't address this issue. They don't mention the `CatalogSearchBuilder`, and says "Most of the SearchBuilders have ::SearchBuilder as a parent or ancestor. ::SearchBuilder does not exist in any repo: it is generated by Blacklight and modified by Hyrax."  This is not true of Hyrax::CatalogSearchBuilder though, which is `Hyrax::CatalogSearchBuilder < Hyrax::SearchBuilder < Blacklight::SearchBuilder < Object`. It does seem to be true of lots of the other search builders -- I didn't realize hyrax had so many of them, and don't entirely understand what they are used for. 

So this is a lack of parallelism in the Hydra architecture with search builders, that makes things confusing -- the catalog one, which has a prime use case for customization, seems to be special, in a way that makes it hard to customize. 

I am also not sure why Sufia/Hyrax deviated from the expected Blacklight architecture with the CatalogSearchBuilder.  It makes things confusing becuase a developer finds the local ::SearchBuilder class, but making changes to it does not effect the CatalogController. Which is ironic, because the CatalogController is actually something that _does_ exist in Blacklight, but has departed from Blacklight architecture the _most_! 

Advice/feedback welcome. I'm not sure how much of this is me not understanding the architecture and it's intent, how much of it is a mistake, or something that wasn't a mistake at the time but is in retrospect, or what. 

@atz

**Comment by jrochkind** [comment]
Actually I'm not sure it's true that all/most Hyrax search builders inherit from locally-generated `::SearchBuilder`. 

[Here's one that does](https://github.com/projecthydra-labs/hyrax/blob/a9e343cc6b4492bf5680647a2133bb1f6190be7c/app/search_builders/hyrax/admin_set_member_search_builder.rb)  `AdminSetMemberSearchBuilder < ::SearchBuilder < (usually unless you mess it up locally) < Blacklight::SearchBuilder < Object`. 

[But here's one that does not](https://github.com/projecthydra-labs/hyrax/blob/a9e343cc6b4492bf5680647a2133bb1f6190be7c/app/search_builders/hyrax/my_shares_search_builder.rb) `Hyrax::MySharesSearchBuilder < Hyrax::SearchBuilder < Blacklight::SearchBuilder < Object`. 

So the existing architecture in Hyrax is inconsistent.

**Comment by jrochkind** [comment]
Here's one hacky workaround to get blacklight_range_limit in Sufia.  Same would work in Hyrax, might be a better way to do it in Hyrax... but I think maybe not. 

```ruby
# ./config/initializers/sufia_catalog_search_builder_overrides.rb

# Unclear how to customize the Sufia::CatalogSearchBuilder, which
# we need to do for blacklight_range_limit.
# https://github.com/projecthydra-labs/hyrax/issues/707

# to_prepare used in case there's dev-mode class reloading of the
# thing we're patching. But we try to check to not do it multiple
# times in case there isn't.
Rails.application.config.to_prepare do
  # would like to look up the class dynamically in case it changes,
  # Sufia does this weird making this the only way, Hyrax is somewhat
  # less weird.
  klass = CatalogController.new.search_builder_class
  unless klass.ancestors.include? BlacklightRangeLimit::RangeLimitBuilder
    klass.send(:include, BlacklightRangeLimit::RangeLimitBuilder)
  end
end
```

I still don't actually have blacklight_range_limit working in my Sufia app, but I'm getting closer, it's doing more than it was before, and displaying all range_limit UI elements (which it wasn't before), but the histogram element is broken. One step at a time!

**Comment by atz** [comment]
Totally agree that the main `Hyrax::CatalogSearchBuilder` is likely to be targeted first for customization by users.  Needs to support that use or provide well-documented alternative.

**Comment by jrochkind** [comment]
A further tangle. I successfully added it to the CatalogController using above logic. 

However, there are other 'backend' places where logged in users get facets. For instance, the "My Works" list. I'm not sure if there are others. These don't use the `Sufia/Hyrax::Hyrax::CatalogSearchBuilder` , but use their own search builders. But somehow inherit the _facet_ config from the CatalogController/CatalogSearchBuilder?

So My Works does not have a date range facet totally set up. I am not sure if I should try to add it to the MyWorksSearchBuilder too -- but there might be other places too? Or try to add it to the MySearchBuilderBehavor module.  Or just add it to `Hyrax/Sufia::SearchBuilder`, which I _think_ all the sufia/hyrax search builders inherit from... unless they inherit from `::SearchBuilder` instead. 

I've tried to monkey patch it into `Sufia::SearchBuilder` so it theoretically effects everything that isn't already effected by `::SearchBuilder`... but so far, it's not working, need to spend more time debugging to figure out why. 

Should I take the silence on this issue to mean that there may be nobody who understands the actual intent of this architecture (or there may actually _be_ no actual intent here), or just that people are busy and/or haven't noticed this?

It would seem to make sense to me that all the Hyrax search builders should have a consistently paralellel superclass-- but I'm nto really sure what should be (obvious choices are `::SearchBuilder` or `Hyrax::SearchBuilder`, since those seem to be the two in effect), or what the implications or potential problems of that would be -- if there's any reason they have inconsistent superclasses to begin with.

**Comment by mjgiarlo** [comment]
@jrochkind :speech_balloon: 
> Should I take the silence on this issue to mean that there may be nobody who understands the actual intent of this architecture (or there may actually be no actual intent here), or just that people are busy and/or haven't noticed this?

The latter, for me anyway.  You may want to add this to Wednesday's Hydra Tech call -- those calls are a good opportunity to flag issues that have gone silent and get synchronous feedback.

**Comment by jrochkind** [comment]
@mjgiarlo Not sure if I have time to be on the tech call wednesday, but whether for this week or the future: How do I add something to a tech call? Apologies if this process is documented somewhere that I haven't found at the moment.

**Comment by mjgiarlo** [comment]
@jrochkind You can edit the agenda of the next meeting here: https://wiki.duraspace.org/display/hydra/Notes+from+Meetings+and+Calls

And if there's no agenda for the next meeting, find the agenda from the last meeting and bug the people signed up as moderator & notetaker. :)

**Comment by jrochkind** [comment]
I gave up on getting "My Works" to support the `date_range_limit`, there were too many places I had to touch with ugly monkey-patching (including a route). 

So I had to _remove_ my date facet from "My Works", so it wouldn't show up at all (rather than show up with an ordinary facet term listing instead of date_range_limit), which I did in a localized my_controller that just removed the field on boot, after the superclass had copied it over from CatalogController. https://github.com/chemheritage/chf-sufia/blob/220109c7dd7f0687512732bdaa9f6409625dff36/app/controllers/my_controller.rb

Not sure if that was the right way to do this or if there is a better way.
